Verify a GPG signature on the release image digest from a root of trust within the payload

smarterclayton opened the pull request:
Implement:

1. A payload can contain a root of trust config map that describes GPG public keys and one or more sources of signatures
2. A verifier that checks the root of trust config map if present in the payload and fails if a signature cannot be found for all public keys from any of the stores
3. Change the CVO to default to "don't allow unverified images", block update to images that haven't been verified, and allow an admin to override that with a field in desiredUpdate `allowUnverifiedImages`
4. Report unverified errors in CVO status as well as the verify bit on the history

In CI, the payload would have a signing key public key mixed in to the payload (OKD, OCP CI).  In OCP, we would mix in an official signing public key and the location of one or more stores to read from.

Questions identified during prototyping:

1. The sync loop does nothing when desiredUpdate is failing to retrieve a payload - we need to "under sync" in that case (continue to sync whatever the last payload was) until the payload is retrieved
2. We need to allow additional stores to be specified on a cluster that define where else source can be located from.  That could be a second config map we mix in that is create only, or it could be global config.
3. What failure modes are not covered here?   The upstream stores going down can be handled by background retry and allowing override.  
4. Update isn't calling `oc adm release info --verify`, which it should (prevents the payload from being unreachable or inconsistent).  However, we'd have to wire through the pull secret into the container that runs oc.
5. Is this too paranoid?  Escape hatch of `--force` makes me feel better, especially when we do automated updates. I would rather have all of our testing using a signature rather than only what end users see.
6. Use something besides atomic container image?  Would prefer not to because a) it's reasonably well reviewed, b) there is already red hat detached signature infra that can publish signatures (we have to talk with RCM more).

TODO:

* [x] Clients have to change (oc adm upgrade and web console)
* [x] Get the CI release signer and verifier up, and public signatures to GCS or other public location
* [ ] Additional testing
* [x] Decide whether the payload should consider its own payload (the innate one) as verified implicitly.  Answer is probably not.

smarterclayton commented:
/test e2e-aws

abhinavdahiya commented on pkg/cvo/cvo.go:
```diff
@@ -197,6 +204,20 @@ func (optr *Operator) InitializeFromPayload() error {
+	}
+	if verifier != nil {
+		glog.Infof("Verifying release authenticity: %v", verifier)
+	} else {
```
would it be better if decided to use Reject based on error type vs nil verifier?

smarterclayton replied:
I only did it for maximum paranoia (easier to review the code block here that guarantees that a verifier is set).

abhinavdahiya commented on pkg/verify/verify.go:
```diff
@@ -0,0 +1,469 @@
+// Reject fails always fails verification.
+var Reject Interface = rejectVerifier{}
+
+// LoadFromPayload looks for a config map in the v1 API group within within the provided update
```
dup `within within`

abhinavdahiya commented on pkg/verify/verify.go:
```diff
@@ -0,0 +1,469 @@
+// verifier-public-key-*: One or more GPG public keys in ASCII form that must have signed the
+//                        release image by digest.
+//
+// store-*: A URL (scheme file://, http://, or https://) location that contains signatures. These
```
nit: we should not allow file:// right? there is not way we can make that happen...

smarterclayton replied:
I'll update the note - I allow it for testing both paths and it's possible we could mount it in to the CVO in the future.  Testing only for now.

abhinavdahiya commented on pkg/verify/verify.go:
```diff
@@ -0,0 +1,469 @@
+// if each provided public key has signed the release image digest. The signature may be in any
+// store and the lookup order is internally defined.
+//
+func LoadFromPayload(update *payload.Update) (Interface, error) {
```
nit it would be better if we did not the errors, and use the `github.com/pkg/errors` for wrapping the errors.

abhinavdahiya commented on pkg/verify/verify.go:
```diff
@@ -0,0 +1,469 @@
+// store and the lookup order is internally defined.
+//
+func LoadFromPayload(update *payload.Update) (Interface, error) {
+	transport, err := transport.New(&transport.Config{})
```
nit: this looks like not the right place to initialize the transport

smarterclayton replied:
Why not?  Transport initialization is cheap?

smarterclayton replied:
oh, I see what you mean.  Yes can go in the loop.

abhinavdahiya commented on pkg/verify/verify.go:
```diff
@@ -0,0 +1,469 @@
+				}
+				stores = append(stores, u)
+			default:
+				glog.Warningf("An unexpected key was found in %s (expected store-* or verifier-public-key-*): %s", src, k)
```
nit: include the action that was taken because of this invalid configuration.

abhinavdahiya commented on pkg/verify/verify.go:
```diff
@@ -0,0 +1,469 @@
+
+var validReleaseDigest = regexp.MustCompile(`^[a-zA-Z0-9:]+$`)
+
+type releaseVerifier struct {
```
This needs description.

abhinavdahiya commented on pkg/verify/verify.go:
```diff
@@ -0,0 +1,469 @@
+		path := base + strconv.Itoa(i)
+		data, err := ioutil.ReadFile(path)
+		if err != nil {
+			if os.IsNotExist(err) {
```
nit: this can be moved to before `err != nil` check.

abhinavdahiya commented on pkg/verify/verify.go:
```diff
@@ -0,0 +1,469 @@
+		data, err := ioutil.ReadFile(path)
+		if err != nil {
+			if os.IsNotExist(err) {
+				if i == 1 {
```
why this check?

smarterclayton replied:
not truly necessary, I can remove it now that there are tests

abhinavdahiya commented on pkg/cvo/cvo.go:
```diff
@@ -218,6 +205,18 @@ func (optr *Operator) InitializeFromPayload() error {
 	}
 	optr.verifier = verifier
 
+	// set up the config sync worker
```
can you add reasoning for this change to the commit?

smarterclayton commented:
Simplified down to fix two problems:

1. Don't try to reverify unverified payloads - the fact that local doesn't need to be verified complicates that, and in the future we can implement caching and other niceties around the verifier and have it be invoked before the retriever
2. Remove the negation of UnverifiedImage and weirdness of the signature of RetrievePayload by having it return a struct with the data needed by the caller.

Added another cvo_scenario test that showed different logic.

smarterclayton commented:
/retest

smarterclayton commented:
/retest

abhinavdahiya commented on pkg/cvo/status.go:
```diff
@@ -377,7 +382,7 @@ func (optr *Operator) syncFailingStatus(original *configv1.ClusterVersion, ierr
 	})
 
-	mergeOperatorHistory(config, optr.currentVersion(), now, false)
+	mergeOperatorHistory(config, optr.currentVersion(), false, now, false)
```
why set image no verified when failing ?

smarterclayton replied:
So verified is "level" which false won't reset verified.  We just don't know here, so we don't change the state of the previous history.

smarterclayton commented:
Are we ok with API changes?  If so #293 needs to be merged before this and I need to bump the API.

smarterclayton commented:
Weird, I can't recreate the unit test flake locally.  Trying more

smarterclayton commented:
Pushed a change to tolerate trailing and leading whitespace on config map keys for stores

smarterclayton commented:
https://github.com/openshift/cluster-update-keys is created and CI will be in place shortly

smarterclayton commented:
It is not payload enabled so we can test PRs to it

smarterclayton commented:
Ok, that run passed except for the auth flake against the registry.

If there's no concerns remaining, I think this is ok to merge.  It will be inert until the payload gets an image with a verifier config map.

smarterclayton commented:
/retest

abhinavdahiya commented:
/test e2e-aws-upgrade

smarterclayton commented:
Weird, it should have picked up the change.

/test e2e-aws-upgrade

smarterclayton commented:
Hrm.  The cvo in the payload will be the previous version.  So running upgrade test here only tests whether the new operator works, not whether it blocks upgrade.

Next retry we should see “force: true” in desired update.

smarterclayton commented:
/retest

smarterclayton commented:
/test e2e-aws-upgrade

smarterclayton commented:
/retest

smarterclayton commented:
This has definitely exacerbated the observedGeneration bug, or made it happen almost consistently.

smarterclayton commented:
/retest

smarterclayton commented:
Will rebase on #176 when it merges and then I think this is ready to go (assuming next upgrade goes green)

abhinavdahiya commented:
/test e2e-aws-upgrade
/test e2e-aws

smarterclayton commented:
WTH

/retest

abhinavdahiya commented:
> WTH
> 
> /retest

Looks like there is a merge conflict @smarterclayton

abhinavdahiya commented:
/lgtm

abhinavdahiya commented:
/retest

smarterclayton commented:
Known flake, force merging.
